### PR TITLE
Add new tools.rename() interface

### LINKED ISSUE #8897: [feature] Provide 'tools.rename' utility in the new tools namespace
A new function to rename files/folders that uses current implementation (_robocopy_ for Windows), but will accept the `conanfile` so we can control behavior in the future if needed

```python
def rename(conanfile, src, dst):
   # Mostly copied from current conans.tools.rename
   ...
```

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: Add new `tools.rename()` interface.
Docs: https://github.com/conan-io/docs/pull/2099

- [x] Refer to the issue that supports this Pull Request: closes #8897
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/test/unittests/tools/files/test_rename.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import os
+import pytest
+
+from conan.tools.files import rename
+from conans.test.utils.mocks import ConanFileMock
+from conans.test.utils.tools import TestClient
+
+
+def test_single_patch_file():
```
Comment (memsharded) [inline conans/test/unittests/tools/files/test_rename.py]: This test is not testing the Windows and the robocopy part. ``os.path.isdir(src)``

Please add a Windows specific test that covers this functionality.

File: conan/tools/files/files.py [hunk 2]
```diff
@@ -150,3 +153,31 @@ def _download_file(file_url):
...
+
+def rename(conanfile, src, dst):
+    """
+    rename a file or folder to avoid "Access is denied" error on Windows
+    :param conanfile: conanfile object
+    :param src: Source file or folder
+    :param dst: Destination file or folder
+    :return: None
+    """
+    if os.path.exists(dst):
```
Comment (memsharded) [inline conan/tools/files/files.py:166]:
```suggestion
    # FIXME: This function has been copied from legacy. Needs to fix: which() call and wrap subprocess call.
    if os.path.exists(dst):
```

**Review by memsharded — COMMENTED** [review]
Looks good, please check review and it can be taken out of draft.

File: conan/tools/files/files.py [hunk 3]
```diff
@@ -150,3 +153,32 @@ def _download_file(file_url):
...
+
+    if platform.system() == "Windows" and which("robocopy") and os.path.isdir(src):
+        # /move Moves files and directories, and deletes them from the source after they are copied.
+        # /e Copies subdirectories. Note that this option includes empty directories.
+        # /ndl Specifies that directory names are not to be logged.
+        # /nfl Specifies that file names are not to be logged.
+        process = subprocess.Popen(["robocopy", "/move", "/e", "/ndl", "/nfl", src, dst],
+                                   stdout=subprocess.PIPE)
+        process.communicate()
+        if process.returncode > 7:  # https://ss64.com/nt/robocopy-exit.html
```
Comment (danimtb) [inline conan/tools/files/files.py:177]: Exit codes from 0 to 7 are considered a success

File: conans/client/tools/files.py [hunk 4]
```diff
@@ -436,7 +437,7 @@ def rename(src, dst):
         process = subprocess.Popen(["robocopy", "/move", "/e", "/ndl", "/nfl", src, dst],
                                    stdout=subprocess.PIPE)
         process.communicate()
-        if process.returncode != 1:
+        if process.returncode > 7:  # https://ss64.com/nt/robocopy-exit.html
```
Comment (danimtb) [inline conans/client/tools/files.py:440]: I have fixed the old implementation as well

File: conans/test/unittests/client/tools/files/rename_test.py [hunk 5]
```diff
@@ -64,3 +64,12 @@ def test_rename(self):
             rename(os.path.join("subdir", "1.txt"), "t.txt")
             self.assertTrue(os.path.isfile(os.path.join(tmpdir, "t.txt")))
             self.assertFalse(os.path.isfile(os.path.join(tmpdir, "subdir", "1.txt")))
+
+    def test_rename_empty_folder(self):
```
Comment (danimtb) [inline conans/test/unittests/client/tools/files/rename_test.py:68]: Added this test to the old implementation as this was failing with robocopy but working with `os.rename`

**Review by memsharded — APPROVED** [review]

**Comment by solvingj** [comment]
@SSE4 @memsharded @danimtb 

Linking here for visibility, it seems like this PR probably relates to, and/or duplicate/mirror/supercede/fix the following issues and PR:

https://github.com/conan-io/conan/issues/1693
https://github.com/conan-io/conan/issues/5205
https://github.com/conan-io/conan/issues/6560
https://github.com/conan-io/conan/pull/5639

If so, we can probably close:
https://github.com/conan-io/conan/pull/5639
https://github.com/conan-io/conan/issues/6560

**Pull request merged**